File: src/rankmark/echo.py
"""The echo: a message pressed into the text with no frame around it. The packet
is the model tag, the coded message and a checksum; every carrier word votes
for one packet bit, chosen by the words just before it. Slots come in runs: a
hash of the previous k token ids either opens a new run at the slot it names
(about one word in eight) or continues the run one slot on. Mirrors echo.js;
the two must stay bit-exact."""

from __future__ import annotations

from dataclasses import dataclass

from .channel import bits_to_bytes, bits_to_int, bytes_to_bits, crc16, int_to_bits

TAG_BITS = 3
CRC_BYTES = 2
K = 4
ANCHOR_EVERY = 8
MAX_BYTES = 8


def echo_len(payload_len: int) -> int:
    return TAG_BITS + 8 * (payload_len + CRC_BYTES)


def build_echo(payload: bytes, tag: int) -> list[int]:
    if not 1 <= len(payload) <= MAX_BYTES:
        raise ValueError(f"echo payload must be 1..{MAX_BYTES} bytes")
    c = crc16(payload)
    return int_to_bits(tag, TAG_BITS) + bytes_to_bits(payload) + bytes_to_bits(bytes([(c >> 8) & 0xFF, c & 0xFF]))


def echo_hash(ids: list[int]) -> int:
    h = 2166136261
    for i in ids:
        h = ((h ^ (i & 0xFFFFFFFF)) * 16777619) & 0xFFFFFFFF
    return h


class EchoSlots:
    def __init__(self, n: int) -> None:
        self.n = n
        self.start: int | None = None
        self.offset = 0

    def next(self, prev_ids: list[int]) -> int:
        h = echo_hash(prev_ids)
        if self.start is None or (h >> 8) % ANCHOR_EVERY == 0:
            self.start = (h >> 11) % self.n
            self.offset = 0
        else:
            self.offset += 1
        return (self.start + self.offset) % self.n


@dataclass
class EchoResult:
    valid: bool
    payload: bytes | None
    tag: int
    tag_ok: bool
    votes: list[int]
    min_votes: int
    covered: int
    bits: list[int]


def parse_echo(llrs: list[float], slots: list[int | None], n: int, lens_tag: int | None = None) -> EchoResult:
    total = [0.0] * n
    votes = [0] * n
    for llr, j in zip(llrs, slots):
        if j is None:
            continue
        total[j] += llr
        votes[j] += 1
    bits = [1 if x < 0 else 0 for x in total]
    tag = bits_to_int(bits[:TAG_BITS])
    body = bits_to_bytes(bits[TAG_BITS:])
    payload_len = len(body) - CRC_BYTES
    payload, got = body[:payload_len], body[payload_len:]
    c = crc16(payload)
    valid = payload_len >= 1 and got == bytes([(c >> 8) & 0xFF, c & 0xFF])
    return EchoResult(valid, payload if valid else None, tag, lens_tag is None or tag == lens_tag,
                      votes, min(votes) if votes else 0, sum(1 for v in votes if v), bits)


def echo_lengths() -> list[int]:
    return [echo_len(i + 1) for i in range(MAX_BYTES)]

File: web/engine/echo.js
// The echo: a message pressed into the text with no frame around it. The
// packet is the model tag, the coded message and a checksum; every carrier
// word votes for one packet bit, chosen by the words just before it, so any
// surviving stretch of text still counts and nothing has to be found first.
// Slots come in runs: a hash of the previous k token ids either opens a new
// run at the slot it names (about one word in eight) or continues the run one
// slot on. Runs keep the coverage even; anchors put writer and reader back in
// step after an edit. Mirrors echo.py; the two must stay bit-exact.

import { bitsToBytes, bytesToBits, crc16, intToBits, bitsToInt } from "./bits.js";

export const ECHO = { tagBits: 3, crcBytes: 2, k: 4, anchorEvery: 8, maxBytes: 8 };

export function echoLen(payloadLen) { return ECHO.tagBits + 8 * (payloadLen + ECHO.crcBytes); }

// the packet bits: tag, message bytes, checksum
export function buildEcho(payload, tag) {
  if (!(payload.length >= 1 && payload.length <= ECHO.maxBytes)) throw new Error(`echo payload must be 1..${ECHO.maxBytes} bytes`);
  const c = crc16(payload);
  return [...intToBits(tag, ECHO.tagBits), ...bytesToBits(payload), ...bytesToBits(Uint8Array.of((c >> 8) & 0xff, c & 0xff))];
}

// FNV-1a over token ids, 32-bit, the same arithmetic in both languages
export function echoHash(ids) {
  let h = 2166136261;
  for (const id of ids) h = Math.imul(h ^ (id >>> 0), 16777619) >>> 0;
  return h;
}

// the slot rule, one state per bit stream and packet length. It advances only
// at carriers, so a run of slots is a run of votes: peek() says which slot a
// word would vote on, and commit() takes the step once the word turns out to
// carry. Runs then tile the packet instead of scattering over it.
export class EchoSlots {
  constructor(n) { this.n = n; this.start = null; this.offset = 0; }
  peek(prevIds) {
    const h = echoHash(prevIds);
    const anchor = this.start === null || ((h >>> 8) % ECHO.anchorEvery) === 0;
    const start = anchor ? (h >>> 11) % this.n : this.start;
    const offset = anchor ? 0 : this.offset + 1;
    return { slot: (start + offset) % this.n, start, offset };
  }
  commit(step) { this.start = step.start; this.offset = step.offset; return step.slot; }
  next(prevIds) { return this.commit(this.peek(prevIds)); }
}

// tally the votes: the summed confidence per slot decides each bit; the tag
// and the checksum say whether the packet is real
export function parseEcho(llrs, slots, n, lensTag = null) {
  const sum = new Array(n).fill(0), votes = new Array(n).fill(0);
  for (let i = 0; i < llrs.length; i++) { const j = slots[i]; if (j === null || j === undefined) continue; sum[j] += llrs[i]; votes[j]++; }
  const bits = sum.map(x => (x < 0 ? 1 : 0));
  const tag = bitsToInt(bits.slice(0, ECHO.tagBits));
  const body = bitsToBytes(bits.slice(ECHO.tagBits));
  const payloadLen = body.length - ECHO.crcBytes;
  const payload = body.slice(0, payloadLen), got = body.slice(payloadLen);
  const c = crc16(payload);
  const valid = payloadLen >= 1 && got[0] === ((c >> 8) & 0xff) && got[1] === (c & 0xff);
  return {
    valid, payload: valid ? payload : null, tag, tagOk: lensTag === null || tag === lensTag,
    votes, minVotes: votes.length ? Math.min(...votes) : 0, covered: votes.filter(v => v > 0).length, bits,
  };
}

// the packet lengths a reader tries: one per message length it might carry
export function echoLengths() { return Array.from({ length: ECHO.maxBytes }, (_, i) => echoLen(i + 1)); }

// what the strip draws: the packet in bit order
export function echoLayout(payloadLen) {
  return [
    { kind: "header", start: 0, len: ECHO.tagBits },
    { kind: "payload", start: ECHO.tagBits, len: 8 * payloadLen },
    { kind: "checksum", start: ECHO.tagBits + 8 * payloadLen, len: 8 * ECHO.crcBytes },
  ];
}
